File: Practice/22/Python/task22.py
def fill(i,j):
    d[i][j] = False
    if (d[i+1][j]):
        i1 = i + 1
        j1 = j
        check(i1,j1)
    if (d[i][j+1]):
        i1 = i
        j1 = j + 1
        check(i1,j1)
    if (d[i-1][j]):
        i1 = i - 1
        j1 = j
        check(i1,j1)
    if (d[i][j-1]):
        i1 = i
        j1 = j - 1
        check(i1,j1)

def check(i1,j1):
    if (maze[i1][j1] == "#"):
        None
    elif (maze[i1][j1] == " "):
        fill(i1, j1)
    else:
        exits[maze[i1][j1]] = True

maze = [                             # 25 str * 27 chr
	"####B######################",
    "# #       #   #      #    #",
    "# # # ###### #### ####### #",
    "# # # #       #   #       #",
    "#   # # ######### # ##### #",
    "# # # #   #       #     # #",
    "### ### ### ############# #",
    "# #   #     # #           #",
    "# # #   ####### ###########",
    "# # # #       # #         C",
    "# # ##### ### # # ####### #",
    "# # #     ### # #       # #",
    "#   ##### ### # ######### #",
    "######### ### #           #",
    "# ####### ### #############",
    "A           #   ###   #   #",
    "# ############# ### # # # #",
    "# ###       # # ### # # # #",
    "# ######### # # ### # # # F",
    "#       ### # #     # # # #",
    "# ######### # ##### # # # #",
    "# #######   #       #   # #",
    "# ####### ######### #######",
    "#         #########       #",
    "#######E############D######"
]
d = [[True]*29]*27
for i in range(25):   # 2 последних доп символа каждой строки = False
    d[i][27] = False
    d[i][28] = False
for i in range(29):   # 2 последних доп строки = False
    d[25][i] = False
    d[26][i] = False
exits = {'A': False, 'B': False, 'C': False, 'D': False, 'E': False, 'F': False}
i,j = input("Координаты: ").split(" ")
i = int(i)
j = int(j)
if (i < 0) or (i > 24) or (j < 0) or (j > 26) or (maze[i][j] == "#"):
    print("Неверные координаты")
else:
    if (maze[i][j] >= 'A') and (maze[i][j] <= 'F'):
        exits[maze[i][j]] = True
    fill(i,j)
    print("Доступные выходы: ")
    a = ord('A')
    while (a <= ord('F')):
        if (exits[chr(a)]):
            print(chr(a))
        a += 1
input("Нажмите Enter")

File: Practice/22/C++/task22/task22.cpp
#include <iostream>
#include <map>
#include <string>

using namespace std;

void check(int i, int j);
void fill(int i, int j);

string maze[25] = {   // 25 str * 27 chr
    "####B######################",
    "# #       #   #      #    #",
    "# # # ###### #### ####### #",
    "# # # #       #   #       #",
    "#   # # ######### # ##### #",
    "# # # #   #       #     # #",
    "### ### ### ############# #",
    "# #   #     # #           #",
    "# # #   ####### ###########",
    "# # # #       # #         C",
    "# # ##### ### # # ####### #",
    "# # #     ### # #       # #",
    "#   ##### ### # ######### #",
    "######### ### #           #",
    "# ####### ### #############",
    "A           #   ###   #   #",
    "# ############# ### # # # #",
    "# ###       # # ### # # # #",
    "# ######### # # ### # # # F",
    "#       ### # #     # # # #",
    "# ######### # ##### # # # #",
    "# #######   #       #   # #",
    "# ####### ######### #######",
    "#         #########       #",
    "#######E############D######"
};
bool d[25][27];
map<char, bool> exits;
int y, x;

void fill(int i, int j)
{
    d[i][j] = false;
    if (i > 0) {
        y = i - 1;
        x = j;
        if (d[y][x]) {
            check(y, x);
        }
    }
    if (j < 26) {
        y = i;
        x = j + 1;
        if (d[y][x]) {
            check(y, x);
        }
    }
    if (i < 24) {
        y = i + 1;
        x = j;
        if (d[y][x]) {
            check(y, x);
        }
    }
    if (j > 0) {
        y = i;
        x = j - 1;
        if (d[y][x]) {
            check(y, x);
        }
    }
}

void check(int i, int j)
{
    if (maze[i][j] == '#') {
    }
    else {
        if (maze[i][j] == ' ') {
            fill(i, j);
        }
        else {
            exits[maze[i][j]] = true;
        }
    }
}

int main()
{
	setlocale(LC_ALL, "Russian");
    int i, j;
    string s = "";
    for (char a = 'A'; a <= 'F'; a++) {
        exits[a] = false;
    }
    for (int a = 0; a < 25; a++) {
        for (int b = 0; b < 27; b++) {
            d[a][b] = true;
        }
    }
    cin >> j >> i;
    if ((i < 0) || (i > 24) || (j < 0) || (j > 26)) {
        cout << "Неверный ввод: выход за границы лабиринта";
    }
    else {
        if (maze[i][j] == '#') {
            cout << "Неверный ввод: стена";
        }
        else {
            check(i, j);
            for (char a = 'A'; a <= 'F'; a++) {
                if (exits[a]) {
                    s = s + " " + a;
                }
            }
            if (s == "") {
                cout << "Выхода нет.";
            }
            else {
                cout << "Доступные выходы:" << s;
            }
        }
    }
    return 0;
}